titus-ext/cassandra/src/main/java/com/netflix/titus/ext/cassandra/store/CassandraJobStore.java

```diff
@@ -606,43 +606,62 @@ public Completable deleteTask(Task task) {
 
     @Override
     public Observable<Job<?>> retrieveArchivedJob(String jobId) {
-        return Observable.fromCallable((Callable<Statement>) () -> retrieveArchivedJobStatement.bind(jobId)).flatMap(statement -> execute(statement)
-                .map(resultSet -> {
-                    Row row = resultSet.one();
-                    if (row == null) {
-                        throw JobStoreException.jobDoesNotExist(jobId);
-                    }
-                    String value = row.getString(0);
-                    return (Job<?>) ObjectMappers.readValue(mapper, value, Job.class);
-                }));
+        Observable<Job> action = retrieveEntityById(jobId, Job.class, retrieveArchivedJobStatement)
+                .switchIfEmpty(retrieveEntityById(jobId, Job.class, retrieveActiveJobStatement))
+                .switchIfEmpty(Observable.error(JobStoreException.jobDoesNotExist(jobId)));
+        return (Observable) action;
     }
 
     @Override
     public Observable<Task> retrieveArchivedTasksForJob(String jobId) {
-        return Observable.fromCallable(() -> retrieveArchivedTaskIdsForJobStatement.bind(jobId).setFetchSize(Integer.MAX_VALUE))
-                .flatMap(retrieveActiveTaskIdsForJob -> execute(retrieveActiveTaskIdsForJob).flatMap(taskIdsResultSet -> {
-                    List<String> taskIds = taskIdsResultSet.all().stream().map(row -> row.getString(0)).collect(Collectors.toList());
-                    List<Observable<ResultSet>> observables = taskIds.stream().map(retrieveArchivedTaskStatement::bind).map(this::execute).collect(Collectors.toList());
-                    return Observable.merge(observables, getConcurrencyLimit()).flatMapIterable(tasksResultSet -> tasksResultSet.all().stream()
-                            .map(row -> row.getString(0))
-                            .map(value -> deserializeTask(value))
-                            .collect(Collectors.toList()));
-                }));
+        return retrieveArchivedTasksForJob(jobId, retrieveArchivedTaskIdsForJobStatement, retrieveArchivedTaskStatement)
+                .switchIfEmpty(retrieveArchivedTasksForJob(jobId, retrieveActiveTaskIdsForJobStatement, retrieveActiveTaskStatement));
+    }
+
+    private Observable<Task> retrieveArchivedTasksForJob(String jobId, PreparedStatement taskIdStatement, PreparedStatement taskStatement) {
+        return Observable.fromCallable(() -> taskIdStatement.bind(jobId).setFetchSize(Integer.MAX_VALUE))
+                .flatMap(retrieveActiveTaskIdsForJob ->
```

Comment:
Is the variable name correct? Seems like these could be active or archived depending on the statement provided.

Reply:
I will change this, and add a comment.